scalanet/src/io/iohk/scalanet/peergroup/PeerGroup.scala

```diff
@@ -57,10 +57,18 @@ trait Channel[A, M] {
 }
 
 object Channel {
+  sealed abstract class IdleState
+  case object ReaderIdle extends IdleState
+  case object WriterIdle extends IdleState
+  case object AllIdle extends IdleState
```

Comment:
Just thinking: would it be easier to represent these with two flags?
```scala
case class ChannelIdle(readerIdle: Boolean, writerIdle: Boolean, first: Boolean) extends ChannelEvent[Nothing]
```

Reply:
tbh I am not great fan of boolean flags in public interfaces, and three of them just make me feel the pain of writing more complex logic on top of it. Also `AllIdle` is totally separate event which can be  configured with totally different time out than other two events.  It maybe matter of opiniong, but imo
```
state match {
 ReaderState =>
 WriterState =>
 AllState =>
}
```
is easier on the eyes than:
```
if (raderIdle && writerIdler) {
} else if (readerIdler) {
} else {
}
```
especially if we were to add `first` flag to the mix.

Reply:
Well you can pattern match on 
```scala
state match {
  case all @ ChannelIdle(true, true, _) =>
  case read @ ChannelIdle(true, _, _) =>
  case write @ ChannelIdle(_, true, _) =>
}
```
You an also make custom extractors to match `All` or `Read` or whatever. 

I guess my main problem is that there's no explanation to _why_ `AllIdle` is a "totally separate event with totally different timeout". Why isn't it just the case of read and write both being idle, according to their respective timeouts?

What would this mean for example? all = 1 minute, read = 5 minutes, write = 10 minutes?

Reply:
so having some thing along the line :
` case all @ ChannelIdle(true, true, _)` looks pretty terrible imo with this amount of booleans, and custom extractors looks like work around for just having `AllIdleState`

As for _why_,  as I see it the main use case for `AllIdle` is when some does not care about read timeout and write timeout, and just want to say `if nothing happens on this channel for this amount of time get me this event` and it is pretty intuitive for this case to have `allIdleTimout` config option and corresponding `AllIdleState` event

As for your question, `read = 1 minute, write = 1 minute, all = 120 minutes` would mean that every minute you get `WriteIdle` and `ReadIdle` events , and every 120 minutes you get `AllIdle` events (assuming nothing happens). It is pretty convoluted but maybe you want to only log things for read and write, and if nothing happens for 120min you want to disconnect.

Reply:
The syntax can be worked around both ways, I wouldn't get hung up on that. It's more the mental model that's different.

For me the natural way to think about it was that reads and writes can time out, and `AllIdle` just means both being idle at the same time. 

But your example is enlightening: `ReadIdle` doesn't mean what I thought it did; it doesn't mean that "read is idle, write is not idle", and vice versa `WriteIdle` doesn't mean that reads happened, but writes didn't. These settings mean how frequently a notification event is supposed to be raised, which changes everything. 

In that case if I set `read = 1 minute, write = 5 minutes` then after 7 minutes I'd see `read; read; read; read; read, write, read; read`, so I'd never see a `readIdle=true, writeIdle=true` combination _unless_ I enable a timeout for `AllIdle`. But then I would have to disable read and write timeouts otherwise I'd get mixed signals, or I'd have to have some aggregator to make sense of it, which in itself would be complicated because it would have to listen to the _absence_ of notifications.

If you could an example which highlights the fact that we're talking about independent notification events (partial states) it would make it much clearer.

I agree it's convoluted.